libgadget/drift.c

```diff
@@ -14,24 +14,31 @@
 
 #define MAXHSML 30000.0
 
-
-static void real_drift_particle(int i, inttime_t ti1, const double ddrift);
+static void
+real_drift_particle(int i, inttime_t ti1, const double ddrift, const double random_shift[3]);
 
 /* Updates a single particle to the current drift time*/
 void drift_particle(int i, inttime_t ti1, struct SpinLocks * spin) {
     if(P[i].Ti_drift == ti1) return;
 
+    double random_shift[3] = {0};
     lock_spinlock(i, spin);
     inttime_t ti0 = P[i].Ti_drift;
     if(ti0 != ti1) {
         const double ddrift = get_drift_factor(ti0, ti1);
-        real_drift_particle(i, ti1, ddrift);
+        real_drift_particle(i, ti1, ddrift, random_shift);
 #pragma omp flush
     }
     unlock_spinlock(i, spin);
 }
 
-static void real_drift_particle(int i, inttime_t ti1, const double ddrift)
+/* Drifts an individual particle to time ti1, by a drift factor ddrift.
+ * The final argument is a random shift vector applied uniformly to all particles before periodic wrapping.
+ * The box is periodic, so this does not affect real physics, but it avoids correlated errors
+ * in the tree expansion.
+ * This function also updates the velocity and updates the density according to an adiabatic factor.
+ */
+static void real_drift_particle(int i, inttime_t ti1, const double ddrift, const double random_shift[3])
```

Comment:
The drift function is an unfortunate mix of several physical processes. 

The reason random_shift has to be plumbed into this function is because we update PEANO key in this function (presumably for better cache coherency), and also because we do the BH reposition in this function. 

Could you add a comment that describes random_shift [3], e.g. 

'In certain steps (currently PM steps) we can add a small random_shift to the position displacement to enhance the accuracy of tree force calculation. The shift is a relative shift between the desired shift in the last step and that of the current shift.' 
my language may be a bit loose, but something like this?